packages/@lwc/node-reactions/src/core/traverse.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import { assert, forEach, isUndefined } from '@lwc/shared';
+import { ReactionRecord, QualifyingReactionTypes } from '../types';
+import {
+    getConnectedRecordsForElement,
+    getDisconnectedRecordsForElement,
+    marker,
+} from './reactions';
+import { queueReactionRecord } from './reaction-queue';
+
+/**
+ * Queue qualifying reaction callbacks for a single node.
+ * If the node is a host element, this method will only queue callbacks for the host element.
+ * It does not process the host's shadow tree.
+ */
+function queueReactionsForSingleElement(
+    elm: Element,
+    reactionTypes: QualifyingReactionTypes,
+    reactionQueue: ReactionRecord[]
+) {
+    // Disconnected callback has to be processed before connected callback
+    if (reactionTypes & 2) {
+        const reactionRecords = getDisconnectedRecordsForElement(elm);
+        if (!isUndefined(reactionRecords)) {
+            queueReactionRecord(reactionQueue, reactionRecords);
+        }
+    }
+    if (reactionTypes & 1) {
+        const reactionRecords = getConnectedRecordsForElement(elm);
+        if (!isUndefined(reactionRecords)) {
+            queueReactionRecord(reactionQueue, reactionRecords);
+        }
+    }
+}
+
+const ShadowRootQuerySelectorAll = ShadowRoot.prototype.querySelectorAll;
```

Comment:
`ShadowRoot` doesn't have a `querySelectorAll` attached from its prototype. It is coming from `DocumentFragment`.

Reply:
In synthetic-shadow mode, the querySelectorAll is resolved on the ShadowRoot's prototype
https://github.com/salesforce/lwc/blob/master/packages/@lwc/synthetic-shadow/src/faux-shadow/shadow-root.ts#L561-L590